File: src/app/actions/posts.ts
"use server";

import { headers } from "next/headers";
import type {
  Post,
  PostTag,
  Subspace,
  Tag,
  User,
} from "@/generated/prisma/client";
import { Prisma } from "@/generated/prisma/client";
import {
  AuthorizationError,
  canEditOwnedResource,
  getAuthenticatedActorFromCookieHeader,
  requireAuthenticated,
} from "@/lib/authorization";
import { prisma } from "@/lib/db";

type PostField = "bodyMarkdown" | "id" | "subspaceId" | "tagIds";

type PostWithRelations = Post & {
  author: Pick<User, "email" | "name" | "pictureUrl" | "sub">;
  subspace: Pick<Subspace, "id" | "name" | "slug">;
  tags: (PostTag & {
    tag: Tag;
  })[];
};

export type PostInput = {
  bodyMarkdown?: unknown;
  subspaceId?: unknown;
  tagIds?: unknown;
};

export type SerializedPostTag = {
  createdAt: string;
  id: string;
  name: string;
  slug: string;
  updatedAt: string;
};

export type SerializedPost = {
  author: {
    email: string;
    name: string | null;
    pictureUrl: string | null;
    sub: string;
  };
  authorSub: string;
  bodyMarkdown: string;
  createdAt: string;
  id: string;
  subspace: {
    id: string;
    name: string;
    slug: string;
  };
  subspaceId: string;
  tags: SerializedPostTag[];
  updatedAt: string;
};

export type PostActionResult =
  | {
      ok: true;
      post: SerializedPost;
    }
  | {
      error: string;
      fieldErrors?: Partial<Record<PostField, string>>;
      ok: false;
    };

export type DeletePostActionResult =
  | {
      id: string;
      ok: true;
    }
  | {
      error: string;
      fieldErrors?: Partial<Record<PostField, string>>;
      ok: false;
    };

type ValidatedPostInput = {
  bodyMarkdown: string;
  subspaceId: string;
  tagIds: string[];
};

const MAX_BODY_MARKDOWN_LENGTH = 100_000;
const MAX_TAGS_PER_POST = 24;

const postInclude = {
  author: {
    select: {
      email: true,
      name: true,
      pictureUrl: true,
      sub: true,
    },
  },
  subspace: {
    select: {
      id: true,
      name: true,
      slug: true,
    },
  },
  tags: {
    include: {
      tag: true,
    },
    orderBy: {
      tag: {
        name: "asc",
      },
    },
  },
} satisfies Prisma.PostInclude;

function serializePost(post: PostWithRelations): SerializedPost {
  return {
    author: post.author,
    authorSub: post.authorSub,
    bodyMarkdown: post.bodyMarkdown,
    createdAt: post.createdAt.toISOString(),
    id: post.id,
    subspace: post.subspace,
    subspaceId: post.subspaceId,
    tags: post.tags.map(({ tag }) => ({
      createdAt: tag.createdAt.toISOString(),
      id: tag.id,
      name: tag.name,
      slug: tag.slug,
      updatedAt: tag.updatedAt.toISOString(),
    })),
    updatedAt: post.updatedAt.toISOString(),
  };
}

function getStringValue(value: unknown): string {
  return typeof value === "string" ? value.trim() : "";
}

function getInputValue(input: FormData | PostInput, field: PostField): unknown {
  if (input instanceof FormData) {
    return input.get(field);
  }

  return input[field as keyof PostInput];
}

function getIdValue(value: unknown): string {
  return typeof value === "string" ? value.trim() : "";
}

function getTagIdValues(input: FormData | PostInput): string[] {
  const values =
    input instanceof FormData
      ? [...input.getAll("tagIds"), ...input.getAll("tagIds[]")]
      : [input.tagIds];

  const splitValues = values.flatMap(
    function collectStringValues(value): string[] {
      if (Array.isArray(value)) {
        return value.flatMap(collectStringValues);
      }

      if (typeof value !== "string") {
        return [];
      }

      return value.split(",");
    },
  );

  return Array.from(
    new Set(
      splitValues
        .map((value) => value.trim())
        .filter((value) => value.length > 0),
    ),
  );
}

function validatePostInput(input: FormData | PostInput):
  | { fieldErrors: Partial<Record<PostField, string>>; ok: false }
  | {
      input: ValidatedPostInput;
      ok: true;
    } {
  const fieldErrors: Partial<Record<PostField, string>> = {};
  const bodyMarkdown = getStringValue(getInputValue(input, "bodyMarkdown"));
  const subspaceId = getStringValue(getInputValue(input, "subspaceId"));
  const tagIds = getTagIdValues(input);

  if (!bodyMarkdown) {
    fieldErrors.bodyMarkdown = "Post body is required.";
  } else if (bodyMarkdown.length > MAX_BODY_MARKDOWN_LENGTH) {
    fieldErrors.bodyMarkdown = `Post body must be ${MAX_BODY_MARKDOWN_LENGTH} characters or fewer.`;
  }

  if (!subspaceId) {
    fieldErrors.subspaceId = "Subspace is required.";
  }

  if (tagIds.length > MAX_TAGS_PER_POST) {
    fieldErrors.tagIds = `Posts can have at most ${MAX_TAGS_PER_POST} tags.`;
  }

  if (Object.keys(fieldErrors).length > 0) {
    return {
      fieldErrors,
      ok: false,
    };
  }

  return {
    input: {
      bodyMarkdown,
      subspaceId,
      tagIds,
    },
    ok: true,
  };
}

async function getAuthenticatedActor() {
  const headerStore = await headers();
  const actor = await getAuthenticatedActorFromCookieHeader(
    headerStore.get("cookie"),
  );

  requireAuthenticated(actor);

  return actor;
}

function isRecordNotFoundError(error: unknown): boolean {
  return (
    error instanceof Prisma.PrismaClientKnownRequestError &&
    error.code === "P2025"
  );
}

function isForeignKeyError(error: unknown): boolean {
  return (
    error instanceof Prisma.PrismaClientKnownRequestError &&
    error.code === "P2003"
  );
}

async function validateReferences(
  input: ValidatedPostInput,
): Promise<Partial<Record<PostField, string>>> {
  const [subspace, tags] = await Promise.all([
    prisma.subspace.findUnique({
      select: {
        id: true,
      },
      where: {
        id: input.subspaceId,
      },
    }),
    input.tagIds.length > 0
      ? prisma.tag.findMany({
          select: {
            id: true,
          },
          where: {
            id: {
              in: input.tagIds,
            },
          },
        })
      : Promise.resolve([]),
  ]);

  const fieldErrors: Partial<Record<PostField, string>> = {};

  if (!subspace) {
    fieldErrors.subspaceId = "Subspace was not found.";
  }

  if (tags.length !== input.tagIds.length) {
    fieldErrors.tagIds = "One or more selected tags were not found.";
  }

  return fieldErrors;
}

function toActionError(error: unknown): PostActionResult {
  if (error instanceof AuthorizationError) {
    return {
      error: error.message,
      ok: false,
    };
  }

  if (isRecordNotFoundError(error)) {
    return {
      error: "Post was not found.",
      fieldErrors: {
        id: "Post was not found.",
      },
      ok: false,
    };
  }

  if (isForeignKeyError(error)) {
    return {
      error: "Post references an invalid subspace, author, or tag.",
      ok: false,
    };
  }

  console.error("Post mutation failed", error);
  return {
    error: "Post mutation failed.",
    ok: false,
  };
}

function toDeleteActionError(error: unknown): DeletePostActionResult {
  if (error instanceof AuthorizationError) {
    return {
      error: error.message,
      ok: false,
    };
  }

  if (isRecordNotFoundError(error)) {
    return {
      error: "Post was not found.",
      fieldErrors: {
        id: "Post was not found.",
      },
      ok: false,
    };
  }

  console.error("Post delete failed", error);
  return {
    error: "Post delete failed.",
    ok: false,
  };
}

export async function createPostAction(
  input: FormData | PostInput,
): Promise<PostActionResult> {
  try {
    const actor = await getAuthenticatedActor();
    const validated = validatePostInput(input);

    if (!validated.ok) {
      return {
        error: "Post input is invalid.",
        fieldErrors: validated.fieldErrors,
        ok: false,
      };
    }

    const referenceErrors = await validateReferences(validated.input);

    if (Object.keys(referenceErrors).length > 0) {
      return {
        error: "Post input is invalid.",
        fieldErrors: referenceErrors,
        ok: false,
      };
    }

    const post = await prisma.$transaction(async (tx) => {
      const createdPost = await tx.post.create({
        data: {
          authorSub: actor.sub,
          bodyMarkdown: validated.input.bodyMarkdown,
          subspaceId: validated.input.subspaceId,
        },
        select: {
          id: true,
        },
      });

      if (validated.input.tagIds.length > 0) {
        await tx.postTag.createMany({
          data: validated.input.tagIds.map((tagId) => ({
            postId: createdPost.id,
            tagId,
          })),
        });
      }

      return tx.post.findUniqueOrThrow({
        include: postInclude,
        where: {
          id: createdPost.id,
        },
      });
    });

    return {
      ok: true,
      post: serializePost(post),
    };
  } catch (error) {
    return toActionError(error);
  }
}

export async function updatePostAction(
  id: string,
  input: FormData | PostInput,
): Promise<PostActionResult> {
  try {
    const actor = await getAuthenticatedActor();
    const postId = getIdValue(id);

    if (!postId) {
      return {
        error: "Post id is required.",
        fieldErrors: {
          id: "Post id is required.",
        },
        ok: false,
      };
    }

    const existingPost = await prisma.post.findUnique({
      select: {
        authorSub: true,
      },
      where: {
        id: postId,
      },
    });

    if (!existingPost) {
      return {
        error: "Post was not found.",
        fieldErrors: {
          id: "Post was not found.",
        },
        ok: false,
      };
    }

    if (!canEditOwnedResource(actor, existingPost.authorSub)) {
      throw new AuthorizationError("Post owner or admin access required", 403);
    }

    const validated = validatePostInput(input);

    if (!validated.ok) {
      return {
        error: "Post input is invalid.",
        fieldErrors: validated.fieldErrors,
        ok: false,
      };
    }

    const referenceErrors = await validateReferences(validated.input);

    if (Object.keys(referenceErrors).length > 0) {
      return {
        error: "Post input is invalid.",
        fieldErrors: referenceErrors,
        ok: false,
      };
    }

    const post = await prisma.$transaction(async (tx) => {
      await tx.post.update({
        data: {
          bodyMarkdown: validated.input.bodyMarkdown,
          subspaceId: validated.input.subspaceId,
        },
        where: {
          id: postId,
        },
      });

      await tx.postTag.deleteMany({
        where: {
          postId,
        },
      });

      if (validated.input.tagIds.length > 0) {
        await tx.postTag.createMany({
          data: validated.input.tagIds.map((tagId) => ({
            postId,
            tagId,
          })),
        });
      }

      return tx.post.findUniqueOrThrow({
        include: postInclude,
        where: {
          id: postId,
        },
      });
    });

    return {
      ok: true,
      post: serializePost(post),
    };
  } catch (error) {
    return toActionError(error);
  }
}

export async function deletePostAction(
  id: string | FormData,
): Promise<DeletePostActionResult> {
  try {
    const actor = await getAuthenticatedActor();
    const postId =
      id instanceof FormData ? getIdValue(id.get("id")) : getIdValue(id);

    if (!postId) {
      return {
        error: "Post id is required.",
        fieldErrors: {
          id: "Post id is required.",
        },
        ok: false,
      };
    }

    const existingPost = await prisma.post.findUnique({
      select: {
        authorSub: true,
      },
      where: {
        id: postId,
      },
    });

    if (!existingPost) {
      return {
        error: "Post was not found.",
        fieldErrors: {
          id: "Post was not found.",
        },
        ok: false,
      };
    }

    if (!canEditOwnedResource(actor, existingPost.authorSub)) {
      throw new AuthorizationError("Post owner or admin access required", 403);
    }

    await prisma.post.delete({
      where: {
        id: postId,
      },
    });

    return {
      id: postId,
      ok: true,
    };
  } catch (error) {
    return toDeleteActionError(error);
  }
}

File: src-rs/posts.rs
use std::{collections::HashSet, error::Error, fmt};

use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{patch, post},
    Json, Router,
};
use getrandom::fill as fill_random;
use serde::{Deserialize, Serialize};

use crate::{
    authorization::{require_author_or_admin, require_post_create, Actor, OptionalActor},
    db::{Database, DbError},
    models::{PostWithRelations, Tag},
    state::AppState,
};

const MAX_BODY_MARKDOWN_LENGTH: usize = 100_000;
const MAX_TAGS_PER_POST: usize = 24;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PostInput {
    body_markdown: Option<String>,
    subspace_id: Option<String>,
    tag_ids: Option<TagIdsInput>,
}

#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum TagIdsInput {
    Many(Vec<String>),
    One(String),
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct PostResponse {
    ok: bool,
    post: SerializedPost,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct DeletePostResponse {
    id: String,
    ok: bool,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct ErrorResponse {
    error: &'static str,
    #[serde(skip_serializing_if = "Option::is_none")]
    field_errors: Option<FieldErrors>,
    ok: bool,
}

#[derive(Debug, Default, Serialize)]
#[serde(rename_all = "camelCase")]
struct FieldErrors {
    #[serde(skip_serializing_if = "Option::is_none")]
    body_markdown: Option<&'static str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    id: Option<&'static str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    subspace_id: Option<&'static str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tag_ids: Option<&'static str>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct SerializedPost {
    author: SerializedAuthor,
    author_sub: String,
    body_markdown: String,
    created_at: String,
    id: String,
    subspace: SerializedPostSubspace,
    subspace_id: String,
    tags: Vec<SerializedPostTag>,
    updated_at: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct SerializedAuthor {
    email: String,
    name: Option<String>,
    picture_url: Option<String>,
    sub: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct SerializedPostSubspace {
    id: String,
    name: String,
    slug: String,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct SerializedPostTag {
    created_at: String,
    id: String,
    name: String,
    slug: String,
    updated_at: String,
}

#[derive(Debug)]
struct ValidatedPostInput {
    body_markdown: String,
    subspace_id: String,
    tag_ids: Vec<String>,
}

#[derive(Debug)]
enum PostError {
    Authorization(crate::authorization::AuthorizationError),
    Database(DbError),
    InvalidInput {
        error: &'static str,
        field_errors: FieldErrors,
    },
    InvalidReference {
        field_errors: FieldErrors,
    },
    NotFound,
    Random(getrandom::Error),
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/posts", post(create_post))
        .route("/api/posts/{id}", patch(update_post).delete(delete_post))
}

async fn create_post(
    State(state): State<AppState>,
    OptionalActor(actor): OptionalActor,
    Json(input): Json<PostInput>,
) -> Response {
    match create_post_inner(state.database, actor, input).await {
        Ok(post) => (
            StatusCode::CREATED,
            Json(PostResponse {
                ok: true,
                post: SerializedPost::from(post),
            }),
        )
            .into_response(),
        Err(error) => error.into_response(),
    }
}

async fn update_post(
    State(state): State<AppState>,
    OptionalActor(actor): OptionalActor,
    Path(id): Path<String>,
    Json(input): Json<PostInput>,
) -> Response {
    match update_post_inner(state.database, actor, id, input).await {
        Ok(post) => Json(PostResponse {
            ok: true,
            post: SerializedPost::from(post),
        })
        .into_response(),
        Err(error) => error.into_response(),
    }
}

async fn delete_post(
    State(state): State<AppState>,
    OptionalActor(actor): OptionalActor,
    Path(id): Path<String>,
) -> Response {
    match delete_post_inner(state.database, actor, id).await {
        Ok(id) => Json(DeletePostResponse { id, ok: true }).into_response(),
        Err(error) => error.into_response(),
    }
}

async fn create_post_inner(
    database: Database,
    actor: Option<Actor>,
    input: PostInput,
) -> Result<PostWithRelations, PostError> {
    let actor = require_post_create(actor).map_err(PostError::Authorization)?;
    let input = validate_post_input(input)?;
    validate_references(&database, &input).await?;
    let id = generate_post_id()?;

    database
        .create_post_with_tags(
            &id,
            actor.sub(),
            &input.body_markdown,
            &input.subspace_id,
            &input.tag_ids,
        )
        .await
        .map_err(PostError::from)
}

async fn update_post_inner(
    database: Database,
    actor: Option<Actor>,
    id: String,
    input: PostInput,
) -> Result<PostWithRelations, PostError> {
    let actor = require_post_create(actor).map_err(PostError::Authorization)?;
    let id = trimmed_string(Some(id));

    if id.is_empty() {
        return Err(PostError::InvalidInput {
            error: "Post id is required.",
            field_errors: FieldErrors {
                id: Some("Post id is required."),
                ..FieldErrors::default()
            },
        });
    }

    let Some(author_sub) = database.post_author_sub(&id).await? else {
        return Err(PostError::NotFound);
    };

    require_author_or_admin(Some(actor), &author_sub).map_err(PostError::Authorization)?;
    let input = validate_post_input(input)?;
    validate_references(&database, &input).await?;

    database
        .update_post_with_tags(
            &id,
            &input.body_markdown,
            &input.subspace_id,
            &input.tag_ids,
        )
        .await
        .map_err(PostError::from)?
        .ok_or(PostError::NotFound)
}

async fn delete_post_inner(
    database: Database,
    actor: Option<Actor>,
    id: String,
) -> Result<String, PostError> {
    let actor = require_post_create(actor).map_err(PostError::Authorization)?;
    let id = trimmed_string(Some(id));

    if id.is_empty() {
        return Err(PostError::InvalidInput {
            error: "Post id is required.",
            field_errors: FieldErrors {
                id: Some("Post id is required."),
                ..FieldErrors::default()
            },
        });
    }

    let Some(author_sub) = database.post_author_sub(&id).await? else {
        return Err(PostError::NotFound);
    };

    require_author_or_admin(Some(actor), &author_sub).map_err(PostError::Authorization)?;

    if !database.delete_post(&id).await? {
        return Err(PostError::NotFound);
    }

    Ok(id)
}

async fn validate_references(
    database: &Database,
    input: &ValidatedPostInput,
) -> Result<(), PostError> {
    let mut field_errors = FieldErrors::default();

    if database.subspace_by_id(&input.subspace_id).await?.is_none() {
        field_errors.subspace_id = Some("Subspace was not found.");
    }

    for tag_id in &input.tag_ids {
        if database.tag_by_id(tag_id).await?.is_none() {
            field_errors.tag_ids = Some("One or more selected tags were not found.");
            break;
        }
    }

    if field_errors.has_errors() {
        return Err(PostError::InvalidReference { field_errors });
    }

    Ok(())
}

fn validate_post_input(input: PostInput) -> Result<ValidatedPostInput, PostError> {
    let body_markdown = trimmed_string(input.body_markdown);
    let subspace_id = trimmed_string(input.subspace_id);
    let tag_ids = normalize_tag_ids(input.tag_ids);
    let mut field_errors = FieldErrors::default();

    if body_markdown.is_empty() {
        field_errors.body_markdown = Some("Post body is required.");
    } else if body_markdown.chars().count() > MAX_BODY_MARKDOWN_LENGTH {
        field_errors.body_markdown = Some("Post body must be 100000 characters or fewer.");
    }

    if subspace_id.is_empty() {
        field_errors.subspace_id = Some("Subspace is required.");
    }

    if tag_ids.len() > MAX_TAGS_PER_POST {
        field_errors.tag_ids = Some("Posts can have at most 24 tags.");
    }

    if field_errors.has_errors() {
        return Err(PostError::InvalidInput {
            error: "Post input is invalid.",
            field_errors,
        });
    }

    Ok(ValidatedPostInput {
        body_markdown,
        subspace_id,
        tag_ids,
    })
}

fn normalize_tag_ids(input: Option<TagIdsInput>) -> Vec<String> {
    let Some(input) = input else {
        return Vec::new();
    };
    let raw_values = match input {
        TagIdsInput::Many(values) => values,
        TagIdsInput::One(value) => vec![value],
    };
    let mut seen = HashSet::new();
    let mut normalized = Vec::new();

    for value in raw_values {
        for tag_id in value
            .split(',')
            .map(str::trim)
            .filter(|tag_id| !tag_id.is_empty())
        {
            if seen.insert(tag_id.to_owned()) {
                normalized.push(tag_id.to_owned());
            }
        }
    }

    normalized
}

fn trimmed_string(value: Option<String>) -> String {
    value.unwrap_or_default().trim().to_owned()
}

fn generate_post_id() -> Result<String, PostError> {
    let mut bytes = [0_u8; 16];
    fill_random(&mut bytes)?;
    Ok(format!("post:{}", hex::encode(bytes)))
}

impl FieldErrors {
    fn has_errors(&self) -> bool {
        self.body_markdown.is_some()
            || self.id.is_some()
            || self.subspace_id.is_some()
            || self.tag_ids.is_some()
    }
}

impl From<PostWithRelations> for SerializedPost {
    fn from(post: PostWithRelations) -> Self {
        Self {
            author: SerializedAuthor {
                email: post.author.email,
                name: post.author.name,
                picture_url: post.author.picture_url,
                sub: post.author.sub,
            },
            author_sub: post.post.author_sub,
            body_markdown: post.post.body_markdown,
            created_at: post.post.created_at.to_rfc3339(),
            id: post.post.id,
            subspace: SerializedPostSubspace {
                id: post.subspace.id,
                name: post.subspace.name,
                slug: post.subspace.slug,
            },
            subspace_id: post.post.subspace_id,
            tags: post.tags.into_iter().map(SerializedPostTag::from).collect(),
            updated_at: post.post.updated_at.to_rfc3339(),
        }
    }
}

impl From<Tag> for SerializedPostTag {
    fn from(tag: Tag) -> Self {
        Self {
            created_at: tag.created_at.to_rfc3339(),
            id: tag.id,
            name: tag.name,
            slug: tag.slug,
            updated_at: tag.updated_at.to_rfc3339(),
        }
    }
}

impl From<DbError> for PostError {
    fn from(error: DbError) -> Self {
        Self::Database(error)
    }
}

impl From<getrandom::Error> for PostError {
    fn from(error: getrandom::Error) -> Self {
        Self::Random(error)
    }
}

impl fmt::Display for PostError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Authorization(error) => write!(formatter, "post authorization failed: {error}"),
            Self::Database(error) => write!(formatter, "post database error: {error}"),
            Self::InvalidInput { error, .. } => formatter.write_str(error),
            Self::InvalidReference { .. } => formatter.write_str("Post input is invalid."),
            Self::NotFound => formatter.write_str("Post was not found."),
            Self::Random(error) => write!(formatter, "post id generation failed: {error}"),
        }
    }
}

impl Error for PostError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            Self::Authorization(error) => Some(error),
            Self::Database(error) => Some(error),
            Self::Random(_) => None,
            Self::InvalidInput { .. } | Self::InvalidReference { .. } | Self::NotFound => None,
        }
    }
}

impl IntoResponse for PostError {
    fn into_response(self) -> Response {
        match self {
            Self::Authorization(error) => error.into_response(),
            Self::InvalidInput {
                error,
                field_errors,
            } => post_error_response(StatusCode::BAD_REQUEST, error, Some(field_errors)),
            Self::InvalidReference { field_errors } => post_error_response(
                StatusCode::BAD_REQUEST,
                "Post input is invalid.",
                Some(field_errors),
            ),
            Self::NotFound => post_error_response(
                StatusCode::NOT_FOUND,
                "Post was not found.",
                Some(FieldErrors {
                    id: Some("Post was not found."),
                    ..FieldErrors::default()
                }),
            ),
            error @ Self::Database(_) if matches!(&error, Self::Database(db_error) if db_error.is_foreign_key_violation()) => {
                post_error_response(
                    StatusCode::BAD_REQUEST,
                    "Post references an invalid subspace, author, or tag.",
                    None,
                )
            }
            error @ (Self::Database(_) | Self::Random(_)) => {
                tracing::error!(error = ?error, message = %error, "post endpoint failed");
                post_error_response(
                    StatusCode::INTERNAL_SERVER_ERROR,
                    "Post mutation failed.",
                    None,
                )
            }
        }
    }
}

fn post_error_response(
    status: StatusCode,
    error: &'static str,
    field_errors: Option<FieldErrors>,
) -> Response {
    (
        status,
        Json(ErrorResponse {
            error,
            field_errors,
            ok: false,
        }),
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn validates_body_subspace_and_tag_count() {
        let error = validate_post_input(PostInput {
            body_markdown: Some(" ".to_owned()),
            subspace_id: Some(" ".to_owned()),
            tag_ids: Some(TagIdsInput::Many(Vec::new())),
        })
        .unwrap_err();

        match error {
            PostError::InvalidInput { field_errors, .. } => {
                assert_eq!(field_errors.body_markdown, Some("Post body is required."));
                assert_eq!(field_errors.subspace_id, Some("Subspace is required."));
            }
            other => panic!("expected invalid input, got {other:?}"),
        }

        let many_tags = (0..=MAX_TAGS_PER_POST)
            .map(|index| format!("tag-{index}"))
            .collect();
        let error = validate_post_input(PostInput {
            body_markdown: Some("hello".to_owned()),
            subspace_id: Some("subspace-1".to_owned()),
            tag_ids: Some(TagIdsInput::Many(many_tags)),
        })
        .unwrap_err();

        match error {
            PostError::InvalidInput { field_errors, .. } => {
                assert_eq!(
                    field_errors.tag_ids,
                    Some("Posts can have at most 24 tags.")
                );
            }
            other => panic!("expected invalid input, got {other:?}"),
        }
    }

    #[test]
    fn normalizes_and_deduplicates_tag_ids() {
        assert_eq!(
            normalize_tag_ids(Some(TagIdsInput::Many(vec![
                " tag-1,tag-2 ".to_owned(),
                "tag-1".to_owned(),
                " ".to_owned(),
            ]))),
            vec!["tag-1".to_owned(), "tag-2".to_owned()]
        );
    }
}
